File: Vitis-AI-Quantizer/vai_q_tensorflow/tensorflow/lite/delegates/gpu/cl/kernels/sigmoid.cc
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/lite/delegates/gpu/cl/kernels/sigmoid.h"

#include "absl/strings/str_cat.h"
#include "tensorflow/lite/delegates/gpu/cl/kernels/util.h"
#include "tensorflow/lite/delegates/gpu/cl/precision.h"

namespace tflite {
namespace gpu {
namespace cl {

Sigmoid::Sigmoid(Sigmoid&& operation)
    : ElementwiseOperation(std::move(operation)) {}

Sigmoid& Sigmoid::operator=(Sigmoid&& operation) {
  if (this != &operation) {
    ElementwiseOperation::operator=(std::move(operation));
  }
  return *this;
}

std::string Sigmoid::GetCoreCode(const std::string& src,
                                 const std::string& z_coord,
                                 const std::string& address) const {
  if (definition_.precision != CalculationsPrecision::F32) {
    return absl::StrCat(
        src, ".x = convert_half(native_recip(1.0f + native_exp(convert_float(-",
        src, ".x))));\n", "  ", src,
        ".y = convert_half(native_recip(1.0f + native_exp(convert_float(-", src,
        ".y))));\n", "  ", src,
        ".z = convert_half(native_recip(1.0f + native_exp(convert_float(-", src,
        ".z))));\n", "  ", src,
        ".w = convert_half(native_recip(1.0f + native_exp(convert_float(-", src,
        ".w))));\n");
  } else {
    return absl::StrCat(src, " = (FLT4)(1.0f) / ((FLT4)(1.0f) + exp(-(", src,
                        ")));\n");
  }
}

Sigmoid CreateSigmoid(const OperationDef& definition) {
  Sigmoid operation(definition);
  operation.SetLinkIndex(0);
  return operation;
}

}  // namespace cl
}  // namespace gpu
}  // namespace tflite

File: src/vai_quantizer/rnn_quantizer/pytorch_binding/pytorch_nndct/nn/modules/sigmoid.py


#
# Copyright 2019 Xilinx Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import torch
import numpy as np
from nndct_shared.quantization import maybe_get_quantizer, process_inputs_and_params, post_quant_process
from nndct_shared.utils import NndctOption
from nndct_shared.base import GLOBAL_MAP, NNDCT_KEYS
from .sigmoid_table import *
from .fix_ops import NndctSigmoidTableLookup, NndctSigmoidSimulation
import pytorch_nndct.utils as py_utils

__all__ = ['Sigmoid']

SIGMOID_TABLE = deephi_sigmoid_table()

class deephi_Sigmoid(torch.nn.modules.Sigmoid):
  r"""DeePhi Sigmoid operation"""

  def __init__(self):
    super(deephi_Sigmoid, self).__init__()
    self.quant_mode, self.quantizer = maybe_get_quantizer()
    self.node = None

  def forward(self, input):

    [input], _ = process_inputs_and_params(
        self.node,
        self.quantizer,
        inputs=[input])

    if NndctOption.nndct_quant_off.value or NndctOption.nndct_cv_app.value:
      output = super().forward(input)
      # quantize output
      [output] = post_quant_process(self.node, [output])
    elif self.quant_mode > 0:
      output = torch.empty_like(input)
      if NndctOption.nndct_tanh_sigmoid_sim.value > 0:
        NndctSigmoidSimulation(input, output)
        [output] = post_quant_process(self.node, [output])
      else:
        input_name = self.node.in_nodes[0]
        fragpos = self.quantizer.get_bnfp(input_name, False)[1]
        quant_device = GLOBAL_MAP.get_ele(NNDCT_KEYS.QUANT_DEVICE)
        Ttable = SIGMOID_TABLE.table.to(quant_device)
        output = output.to(quant_device)
        NndctSigmoidTableLookup(input,
                                Ttable,
                                output,
                                fragpos)
    else:
      output = super().forward(input)

    return output


@py_utils.register_quant_op
def Sigmoid(*args, **kwargs):
  quant_mode, _ = maybe_get_quantizer()
  if quant_mode == None:
    return torch.nn.Sigmoid(*args, **kwargs)
  return deephi_Sigmoid(*args, **kwargs)
